packages/coreutils/src/poll.ts

```diff
@@ -0,0 +1,657 @@
+   */
+  async stop(): Promise<void> {
+    if (this.isDisposed) {
+      return;
+    }
+
+    if (this.state.phase === 'constructed') {
+      await this.ready;
+    }
+
+    if (this.state.phase !== 'stopped') {
+      this.schedule({
+        interval: Private.NEVER,
+        payload: null,
+        phase: 'stopped',
+        timestamp: new Date().getTime()
+      });
+    }
+  }
+
+  /**
+   * A promise that resolves after the poll has scheduled its first tick.
+   *
+   * #### Notes
+   * This accessor is protected to allow sub-classes to implement methods that
+   * can `await this.ready` if `this.state.phase === 'constructed'`.
+   *
+   * A poll should handle ready state without needing to expose it to clients.
+   */
+  protected readonly ready: Promise<void>;
+
+  /**
+   * Schedule the next poll tick.
+   *
+   * @param tick - The new tick data to populate the poll state.
+   *
+   * #### Notes
+   * This method is protected to allow sub-classes to implement methods that can
+   * schedule poll ticks.
+   *
+   * This method synchronously modifies poll state so it should be invoked with
+   * consideration given to the poll state that will be overwritten. It will
+   * typically be invoked in a method that returns a promise, allowing the
+   * caller e.g. to `await this.ready` before scheduling a new tick.
+   */
+  protected schedule(tick: IPoll.Tick<T, U>): void {
+    const current = new PromiseDelegate<this>();
+    const pending = this._tick;
+    const execute = () => {
+      if (this.isDisposed || this.tick !== current.promise) {
+        return;
+      }
+
+      this._execute();
+    };
+
+    // Clear the schedule if possible.
+    if (this.state.interval === Private.IMMEDIATE) {
+      cancelAnimationFrame(this._timeout);
+    } else {
+      clearTimeout(this._timeout);
+    }
+
+    // Update poll state and schedule the next tick.
+    this._state = tick;
+    this._tick = current;
+    this._timeout =
+      tick.interval === Private.IMMEDIATE
+        ? requestAnimationFrame(execute)
+        : tick.interval === Private.NEVER
+          ? -1
+          : setTimeout(execute, tick.interval);
+
+    // Resolve the pending poll promise and emit the ticked signal.
+    void pending.promise.then(() => {
+      this._ticked.emit(tick);
+    });
+    pending.resolve(this);
+  }
+
+  /**
+   * Execute a new poll factory promise or stand by if necessary.
+   */
+  private _execute(): void {
+    let standby =
+      typeof this.standby === 'function' ? this.standby() : this.standby;
+    standby =
+      standby === 'never'
+        ? false
+        : standby === 'when-hidden'
+          ? !!(typeof document !== 'undefined' && document && document.hidden)
+          : true;
+
+    // If in standby mode schedule next tick without calling the factory.
+    if (standby) {
+      this.schedule({
+        interval: this.frequency.interval,
+        payload: null,
+        phase: 'standby',
+        timestamp: new Date().getTime()
```

Comment:
We are constantly creating these timestamps. Perhaps it would be easier if `schedule` created them automatically for us?

Reply:
Yeah that is not a bad idea, I just didn't want to pass `Partial<Tick>` objects around.

Reply:
I was thinking the argument for `schedule` would be something like
```typescript
Pick<IPoll.Tick<T, U>, Exclude<keyof IPoll.Tick<T, U>, 'timestamp'>>
```

Reply:
Maybe it's more confusing than anything. And perhaps it's useful for a user to be able to set the timestamp to be able to distinguish ticks with something they set.